src/client.ts

```diff
@@ -76,7 +78,128 @@ export default function connect(
   dsn?: string | ConnectConfig | null,
   options?: ConnectConfig | null
 ): Promise<Connection> {
-  return ConnectionImpl.connect(dsn, options);
+  let config: ConnectConfig | null = null;
+  if (typeof dsn === "string") {
+    config = {...options, dsn};
+  } else {
+    if (dsn != null) {
+      // tslint:disable-next-line: no-console
+      console.warn(
+        "`options` as the first argument to `edgedb.connect` is " +
+          "deprecated, use " +
+          "`edgedb.connect('instance_name_or_dsn', options)`"
+      );
+    }
+    config = {...dsn, ...options};
+  }
+  return StandaloneConnection.connect(parseConnectArguments(config));
+}
+
+function sleep(durationMicros: number): Promise<void> {
+  return new Promise((accept, reject) => {
+    setTimeout(() => accept(), durationMicros);
+  });
+}
+
+class StandaloneConnection implements Connection {
+  [ALLOW_MODIFICATIONS]: never;
+  private config: NormalizedConnectConfig;
+  private _connection?: ConnectionImpl;
+  private _isClosed: boolean; // For compatibility
+
+  private constructor(config: NormalizedConnectConfig) {
+    this.config = config;
+    this._isClosed = false;
+  }
+  async transaction<T>(
+    action: () => Promise<T>,
+    options?: TransactionOptions
+  ): Promise<T> {
+    return await (await this.connection()).transaction(action, options);
+  }
+  async close(): Promise<void> {
+    try {
+      if (this._connection) {
+        await this._connection.close();
+      }
+      // TODO(tailhook) it makes little sense to close the reconnecting
+      // connection so maybe deprecate this method
+      this._isClosed = true;
+    } finally {
+      this._cleanupProxy();
+    }
+  }
+  private _cleanupProxy(): void {
+    const proxy = proxyMap.get(this);
+    if (proxy != null) {
+      proxy[onConnectionClose]();
+      proxyMap.delete(this);
+    }
+  }
+  isClosed(): boolean {
+    return this._isClosed;
+  }
+  async execute(query: string): Promise<void> {
+    return await (await this.connection()).execute(query);
```

Comment:
I'd like to avoid having one extra await here (for perf reasons) and replace this with something like this, in pseudocode:

```
let connection: Connection | null = this._tryGetConnection();
if (connection == null) {
  connection = await this._getConnection();
}
```

Reply:
Well, can't V8 do it for us? If we're going to inline functions manually, why don't we inline the `execute` function too, as it's kinda about the same number of lines.
```
  async execute(query: string): Promise<void> {
    this._enterOp();
    try {
      await this._execute(query);
    } finally {
      this._leaveOp();
    }
  }

```
Do we have any estimation of how costly awaits are? Especially since `await` you're refering to is not nested comparing to `await` in my example which contributes to the call stack.

Reply:
> Well, can't V8 do it for us? If we're going to inline functions manually, why don't we inline the execute function too, as it's kinda about the same number of lines.

I don't think so. async/await isn't yet popular enough to deserve a lot of attention from V8 engineers. And JS isn't Rust; optimizations like this are somewhat more complex when implemented in a JIT. But we're talking about just 5-6 functions here, inlining by hand is easy.

Reply:
Done.